backend/tournesol/views/polls.py

```diff
@@ -149,6 +149,13 @@ def filter_unsafe(self, queryset, filters):
             tournesol_score__gt=0,
         )
 
+    def exclude_rated_entities(self, queryset, filters, user):
+        if filters["exclude_rated_entities"] and user.is_authenticated:
+            user_rating_qs = ContributorRating.objects.filter(user=user)
```

Comment:
I think I found a possible bug here.

`ContributorRating.objects.filter(user=user)` is an approximation of the entities rated by the user, it doesn't represent exactly the rated entities.

In fact it's possible for a user to create one or more `ContributorRating` objects without comparing the related entities. This is what happens when we paste YouTube URLs in the comparison page. When receiving the URL,  and even before the comparison is submitted, the front end component creates a  `ContributorRating` for the couple user / entities.

**option (a)**

A better approximation could be (SQL pseudo code):
- SELECT the `ContributorRating`
- INNER JOIN `ContributorRatingCriteriaScore`
- ON `ContributorRatingCriteriaScore.contributor_rating_id` = `ContributorRating.id`
- WHERE `ContributorRating.user_id` = authenticated user id
- AND `ContributorRating.poll_id` = poll.id
- AND `ContributorRatingCriteriaScore.criteria` = poll.main_criteria

Here we are doing a SQL INNER JOIN to keep only the compared entities. The compared entities are the entities having at least a computed score for the main criteria (largely recommended for the `videos` poll).

This is still an approximation as the score are computed only after a run of the algorithm, meaning it's possible to compare an entity and to have the `ContributorRatingCriteriaScore` table empty (until the next run).

**option (b)**

If we want to exclude the compared entities in the real time, we need to filter on the rating given by the user at the time of the comparison instead of the computed score which exists only after a certain amount of time in the database.

SQL pseudo code:

- SELECT the Comparison
- JOIN  ComparisonCriteriaScore
- ON ComparisonCriteriaScore.comparison_id = Comparison.id
- WHERE Comparison.user_id = authenticated user id
- AND Comparison.poll_id = poll.id
- AND ComparisonCriteriaScore.criteria = poll.main_criteria

If an entity has been compared multiple times by the user, it will appear several times in the results of the previous SQL query. I don't know how it will impact the performance of global recommendations queryset, and if it's relevant to filter the duplicates. If it is we either need to filter the duplicates in the SQL query (might be tricky), or to filter them directly in Python with `set` (may be slower than a filter in the SQL query).

I propose to investigate the feasibility of the **option (b)** first, as it represents what we want to have minus the duplicates. We can check the performance of option b:
- with the duplicates
- without the duplicates

Reply:
> SQL pseudo code:
> 
> * SELECT the Comparison
> * JOIN  ComparisonCriteriaScore
> * ON ComparisonCriteriaScore.comparison_id = Comparison.id
> * WHERE Comparison.user_id = authenticated user id
> * AND Comparison.poll_id = poll.id
> * AND ComparisonCriteriaScore.criteria = poll.main_criteria

I don't understand why do you recommend to use the ComparisonCriteriaScore too? Can there be comparison object without the associated criteria scores?

Reply:
It looks unnecessary indeed, it's possibly a leftover of the option (a) pseudo query.

Currently our API doesn't allow users to create comparison without the main criterion.

(Nevertheless, I remember correctly we have comparisons in the database that don't have a related `ComparisonCriteriaScore` for the main criterion, but the suggested SQL join won't fix the issue).